Ladybird/LocationEdit.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2023, Cameron Youell <cameronyouell@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "LocationEdit.h"
+#include "Tab.h"
+#include <AK/URL.h>
+#include <QCoreApplication>
+#include <QTextLayout>
+
+LocationEdit::LocationEdit(Tab* tab)
+    : QLineEdit(tab)
+    , m_tab(tab)
+{
+    connect(this, &QLineEdit::returnPressed, this, [&] {
+        clearFocus();
+    });
+
+    connect(this, &QLineEdit::textChanged, this, [&] {
+        highlightLocation();
+    });
+}
+
+void LocationEdit::focusInEvent(QFocusEvent* event)
+{
+    QLineEdit::focusInEvent(event);
+    highlightLocation();
+}
+
+void LocationEdit::focusOutEvent(QFocusEvent* event)
+{
+    QLineEdit::focusOutEvent(event);
+    highlightLocation();
+}
+
+void LocationEdit::highlightLocation()
+{
+    AK::URL url = AK::URL::create_with_url_or_path(text().toStdString().c_str());
+
+    QList<QInputMethodEvent::Attribute> attributes;
+    if (url.is_valid() && !hasFocus()) {
+        if (url.scheme() == "http" || url.scheme() == "https" || url.scheme() == "gemini") {
+            int host_start = (url.scheme().length() + 3) - cursorPosition();
+            auto host_length = url.host().length();
+
+            if (auto host_parts = url.host().split('.'); host_parts.size() > 2) {
+                host_length = 0;
+                for (auto i = host_parts.size() - 2; i < host_parts.size(); i++) {
+                    host_length += host_parts[i].length() + 1;
+                }
+                host_start += url.host().length() - host_length;
+
+                host_start++;
+                host_length--;
+            }
+
+            QTextCharFormat defaultFormat;
+            defaultFormat.setForeground(Qt::darkGray);
```

Comment:
Don't hardcode UI colors, some of us are using dark themes

Reply:
Fixed :^)